lib/dat-worker-pool/worker.rb

```diff
@@ -100,42 +100,61 @@ def work!(work_item)
       end
 
       # rescue `ShutdownError` and ignore it, its already been passed to the
-      # on-error callbacks in `fetch_and_work`; the shutdown error is just used
-      # to force the worker to exit its infinite loop if the worker pool is
+      # on-error callbacks in `dwp_fetch_and_work`; the shutdown error is just
+      # used to force the worker to exit its infinite loop if the worker pool is
       # forcing itself to shutdown
-      def work_loop
-        run_callback 'on_start'
-        fetch_and_work while self.running?
+      def dwp_work_loop
+        dwp_setup
+        dwp_fetch_and_work while self.running?
       rescue ShutdownError
       ensure
-        run_callback 'on_shutdown'
-        @runner.despawn_worker(self)
-        @thread = nil
+        dwp_teardown
+      end
+
+      def dwp_setup
+        @dwp_runner.spawn_worker(self)
+        begin
+          dwp_run_callback 'on_start'
+        rescue StandardError, ShutdownError => exception
+          dwp_handle_exception(exception)
+          Thread.current.raise exception
+        end
+      end
+
+      def dwp_teardown
+        begin
+          dwp_run_callback 'on_shutdown'
+        rescue StandardError, ShutdownError => exception
```

Comment:
@jcredding we discussed only handling shutdown error exceptions if a work item is in play.